Source/bmalloc/libpas/src/test/EnumerationTests.cpp

```diff
@@ -0,0 +1,315 @@
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <map>
+#include <stdlib.h>
+#include <set>
+#include <sys/mman.h>
+
+#include "TestHarness.h"
+#include "pas_root.h"
+#include "bmalloc_heap.h"
+#include "pas_probabilistic_guard_malloc_allocator.h"
+#include "pas_heap.h"
+#include "iso_heap.h"
+#include "iso_heap_config.h"
+#include "pas_heap_ref_kind.h"
+
+
+using namespace std;
+
+namespace {
+
+const bool verbose = false;
+
+struct PageRange {
+    PageRange() = default;
+
+    PageRange(void* base, size_t size)
+        : base(base)
+        , size(size)
+    {
+        PAS_ASSERT(pas_is_aligned(reinterpret_cast<uintptr_t>(base), pas_page_malloc_alignment()));
+        PAS_ASSERT(pas_is_aligned(size, pas_page_malloc_alignment()));
+        PAS_ASSERT(!!base == !!size);
+    }
+
+    bool operator<(PageRange range) const
+    {
+        return base < range.base;
+    }
+
+    void* end() const
+    {
+        return reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(base) + size);
+    }
+
+    void* base { nullptr };
+    size_t size { 0 };
+};
+
+set<PageRange> pageRanges;
+
+struct RecordedRange {
+    RecordedRange() = default;
+
+    RecordedRange(void* base, size_t size)
+        : base(base)
+        , size(size)
+    {
+        PAS_ASSERT(base);
+    }
+
+    bool operator<(RecordedRange other) const
+    {
+        return base < other.base;
+    }
+
+    void* end() const
+    {
+        return reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(base) + size);
+    }
+
+    void* base { nullptr };
+    size_t size { 0 };
+};
+
+struct ReaderRange {
+    ReaderRange() = default;
+
+    ReaderRange(void* base, size_t size)
+        : base(base)
+        , size(size)
+    {
+        PAS_ASSERT(base);
+        PAS_ASSERT(size);
+    }
+
+    bool operator<(ReaderRange other) const
+    {
+        if (base != other.base)
+            return base < other.base;
+        return size < other.size;
+    }
+
+    void* base { nullptr };
+    size_t size { 0 };
+};
+
+map<pas_enumerator_record_kind, set<RecordedRange>> recordedRanges;
+map<ReaderRange, void*> readerCache;
+
+void* enumeratorReader(pas_enumerator* enumerator,
+                       void* address,
+                       size_t size,
+                       void* arg)
```

Comment:
Nit: We don't generally declare function arguments with each on their own line.

Reply:
This is code taken from another test case that Phil wrote.

I changed it to one line.